File: compat/suites/dotnet-sdk/Groups/SqsGroup.cs
using Amazon.SQS.Model;
using OvercastCompat.Clients;
using OvercastCompat.Harness;

namespace OvercastCompat.Groups;

public sealed class SqsGroup(AwsClients clients) : IServiceGroup
{
    public IReadOnlyDictionary<string, TestFn> Impls() => new Dictionary<string, TestFn>(StringComparer.Ordinal)
    {
        ["sqs-messages:SendMessage"] = SendMessageAsync,
        ["sqs-messages:SendMessageBatch"] = SendMessageBatchAsync,
        ["sqs-messages:ReceiveMessage"] = ReceiveMessageAsync,
        ["sqs-messages:DeleteMessage"] = DeleteMessageAsync,
        ["sqs-messages:ChangeMessageVisibility"] = ChangeMessageVisibilityAsync,
        ["sqs-messages:DeleteMessageBatch"] = DeleteMessageBatchAsync,
        ["sqs-messages:PurgeQueue"] = PurgeQueueAsync,
        ["sqs-dlq:CreateDLQ"] = CreateDLQAsync,
        ["sqs-dlq:SetRedrivePolicy"] = SetRedrivePolicyAsync,
        ["sqs-dlq:GetRedrivePolicy"] = GetRedrivePolicyAsync,
        ["sqs-fifo:CreateFifoQueue"] = CreateFifoQueueAsync,
        ["sqs-fifo:SendFifoMessage"] = SendFifoMessageAsync,
        ["sqs-fifo:ReceiveFifoMessage"] = ReceiveFifoMessageAsync,
    };

    public IReadOnlyDictionary<string, SetupFn> Setups() => new Dictionary<string, SetupFn>(StringComparer.Ordinal)
    {
        ["sqs-messages"] = SetupMessagesAsync,
        ["sqs-dlq"] = SetupDlqAsync,
    };

    public IReadOnlyDictionary<string, SetupFn> Teardowns() => new Dictionary<string, SetupFn>(StringComparer.Ordinal)
    {
        ["sqs-messages"] = TeardownMessagesAsync,
        ["sqs-dlq"] = TeardownDlqAsync,
        ["sqs-fifo"] = TeardownFifoAsync,
    };

    // ── sqs-messages ──

    private async Task SetupMessagesAsync(TestContext context)
    {
        var name = $"{context.RunId}-sqs-msg";
        var response = await clients.SQS().CreateQueueAsync(new CreateQueueRequest { QueueName = name });
        context.Set("sqsMsgUrl", response.QueueUrl);
    }

    private async Task SendMessageAsync(TestContext context)
    {
        var url = RequireQueueUrl(context, "sqsMsgUrl");
        var response = await clients.SQS().SendMessageAsync(new SendMessageRequest
        {
            QueueUrl = url,
            MessageBody = "hello from dotnet-sdk",
        });
        Assertions.NotBlank(response.MessageId, "SendMessage: MessageId");
    }

    private async Task SendMessageBatchAsync(TestContext context)
    {
        var url = RequireQueueUrl(context, "sqsMsgUrl");
        var response = await clients.SQS().SendMessageBatchAsync(new SendMessageBatchRequest
        {
            QueueUrl = url,
            Entries =
            [
                new SendMessageBatchRequestEntry { Id = "a", MessageBody = "batch-a" },
                new SendMessageBatchRequestEntry { Id = "b", MessageBody = "batch-b" },
            ],
        });
        Assertions.GreaterThanOrEqual(2, response.Successful.Count, "SendMessageBatch: expected >= 2 successful");
        Assertions.NotBlank(response.Successful[0].MessageId, "SendMessageBatch: MessageId[0]");
    }

    private async Task ReceiveMessageAsync(TestContext context)
    {
        var url = RequireQueueUrl(context, "sqsMsgUrl");
        await clients.SQS().SendMessageAsync(new SendMessageRequest { QueueUrl = url, MessageBody = "receive-test" });
        List<Message> messages = new();
        for (var i = 0; i < 5; i++)
        {
            var response = await clients.SQS().ReceiveMessageAsync(new ReceiveMessageRequest
            {
                QueueUrl = url,
                MaxNumberOfMessages = 10,
                WaitTimeSeconds = 1,
            });
            messages = response.Messages;
            if (messages.Count > 0)
            {
                break;
            }
        }
        Assertions.True(messages.Count > 0, "ReceiveMessage: expected at least 1 message");
    }

    private async Task DeleteMessageAsync(TestContext context)
    {
        var url = RequireQueueUrl(context, "sqsMsgUrl");
        await clients.SQS().SendMessageAsync(new SendMessageRequest { QueueUrl = url, MessageBody = "delete-me" });
        var recv = await clients.SQS().ReceiveMessageAsync(new ReceiveMessageRequest
        {
            QueueUrl = url,
            MaxNumberOfMessages = 1,
            WaitTimeSeconds = 3,
        });
        Assertions.True(recv.Messages.Count > 0, "DeleteMessage: expected at least 1 message");
        var handle = recv.Messages[0].ReceiptHandle;
        Assertions.NotBlank(handle, "DeleteMessage: ReceiptHandle");
        await clients.SQS().DeleteMessageAsync(new DeleteMessageRequest { QueueUrl = url, ReceiptHandle = handle });
    }

    private async Task ChangeMessageVisibilityAsync(TestContext context)
    {
        var url = RequireQueueUrl(context, "sqsMsgUrl");
        await clients.SQS().SendMessageAsync(new SendMessageRequest { QueueUrl = url, MessageBody = "visibility-test" });
        var recv = await clients.SQS().ReceiveMessageAsync(new ReceiveMessageRequest
        {
            QueueUrl = url,
            MaxNumberOfMessages = 1,
            WaitTimeSeconds = 3,
        });
        Assertions.True(recv.Messages.Count > 0, "ChangeMessageVisibility: expected at least 1 message");
        await clients.SQS().ChangeMessageVisibilityAsync(new ChangeMessageVisibilityRequest
        {
            QueueUrl = url,
            ReceiptHandle = recv.Messages[0].ReceiptHandle,
            VisibilityTimeout = 30,
        });
    }

    private async Task DeleteMessageBatchAsync(TestContext context)
    {
        var url = RequireQueueUrl(context, "sqsMsgUrl");
        await clients.SQS().SendMessageAsync(new SendMessageRequest { QueueUrl = url, MessageBody = "batch-delete-a" });
        await clients.SQS().SendMessageAsync(new SendMessageRequest { QueueUrl = url, MessageBody = "batch-delete-b" });
        var recv = await clients.SQS().ReceiveMessageAsync(new ReceiveMessageRequest
        {
            QueueUrl = url,
            MaxNumberOfMessages = 10,
            WaitTimeSeconds = 3,
        });
        Assertions.GreaterThanOrEqual(2, recv.Messages.Count, "DeleteMessageBatch: expected >= 2 messages");
        var batchResp = await clients.SQS().DeleteMessageBatchAsync(new DeleteMessageBatchRequest
        {
            QueueUrl = url,
            Entries = recv.Messages.Select((msg, idx) => new DeleteMessageBatchRequestEntry
            {
                Id = (idx + 1).ToString(),
                ReceiptHandle = msg.ReceiptHandle,
            }).ToList(),
        });
        Assertions.GreaterThanOrEqual(2, batchResp.Successful.Count, "DeleteMessageBatch: expected >= 2 successful deletes");
    }

    private async Task PurgeQueueAsync(TestContext context)
    {
        var url = RequireQueueUrl(context, "sqsMsgUrl");
        await clients.SQS().SendMessageAsync(new SendMessageRequest { QueueUrl = url, MessageBody = "purge-test" });
        await clients.SQS().PurgeQueueAsync(new PurgeQueueRequest { QueueUrl = url });
        var recv = await clients.SQS().ReceiveMessageAsync(new ReceiveMessageRequest
        {
            QueueUrl = url,
            MaxNumberOfMessages = 10,
            WaitTimeSeconds = 0,
        });
        // The AWS SDK for .NET v4 leaves response collections null when the
        // service returns none, so ".Count" throws rather than reading 0.
        Assertions.Equal(0, recv.Messages?.Count ?? 0, "PurgeQueue: expected no messages after purge");
    }

    private async Task TeardownMessagesAsync(TestContext context)
    {
        var url = context.GetString("sqsMsgUrl");
        if (!string.IsNullOrWhiteSpace(url))
        {
            try { await clients.SQS().DeleteQueueAsync(new DeleteQueueRequest { QueueUrl = url }); } catch { }
        }
    }

    // ── sqs-dlq ──

    private async Task SetupDlqAsync(TestContext context)
    {
        var srcName = $"{context.RunId}-sqs-src";
        var dlqName = $"{context.RunId}-sqs-dlq";
        var srcResponse = await clients.SQS().CreateQueueAsync(new CreateQueueRequest { QueueName = srcName });
        var dlqResponse = await clients.SQS().CreateQueueAsync(new CreateQueueRequest { QueueName = dlqName });
        context.Set("sqsSrcUrl", srcResponse.QueueUrl);
        context.Set("sqsDlqUrl", dlqResponse.QueueUrl);
    }

    private async Task CreateDLQAsync(TestContext context)
    {
        var srcUrl = context.GetString("sqsSrcUrl") ?? throw new InvalidOperationException("sqsSrcUrl not set");
        var dlqUrl = context.GetString("sqsDlqUrl") ?? throw new InvalidOperationException("sqsDlqUrl not set");
        var list = await clients.SQS().ListQueuesAsync(new ListQueuesRequest { QueueNamePrefix = $"{context.RunId}-sqs-src" });
        Assertions.True(list.QueueUrls.Any(u => u == srcUrl), $"CreateDLQ: src queue not found (runId={context.RunId})");
        var listDlq = await clients.SQS().ListQueuesAsync(new ListQueuesRequest { QueueNamePrefix = $"{context.RunId}-sqs-dlq" });
        Assertions.True(listDlq.QueueUrls.Any(u => u == dlqUrl), $"CreateDLQ: dlq queue not found (runId={context.RunId})");
    }

    private async Task SetRedrivePolicyAsync(TestContext context)
    {
        var srcUrl = context.GetString("sqsSrcUrl") ?? throw new InvalidOperationException("sqsSrcUrl not set");
        var dlqUrl = context.GetString("sqsDlqUrl") ?? throw new InvalidOperationException("sqsDlqUrl not set");
        var dlqAttrs = await clients.SQS().GetQueueAttributesAsync(new GetQueueAttributesRequest
        {
            QueueUrl = dlqUrl,
            AttributeNames = ["QueueArn"],
        });
        var dlqArn = dlqAttrs.Attributes["QueueArn"];
        var policy = $"{{\"maxReceiveCount\":\"3\",\"deadLetterTargetArn\":\"{dlqArn}\"}}";
        await clients.SQS().SetQueueAttributesAsync(new SetQueueAttributesRequest
        {
            QueueUrl = srcUrl,
            Attributes = new Dictionary<string, string> { ["RedrivePolicy"] = policy },
        });
        var srcAttrs = await clients.SQS().GetQueueAttributesAsync(new GetQueueAttributesRequest
        {
            QueueUrl = srcUrl,
            AttributeNames = ["RedrivePolicy"],
        });
        Assertions.True(srcAttrs.Attributes.ContainsKey("RedrivePolicy"), "SetRedrivePolicy: RedrivePolicy not found");
        Assertions.True(srcAttrs.Attributes["RedrivePolicy"].Contains("deadLetterTargetArn"), "SetRedrivePolicy: missing deadLetterTargetArn");
    }

    private async Task GetRedrivePolicyAsync(TestContext context)
    {
        var srcUrl = context.GetString("sqsSrcUrl") ?? throw new InvalidOperationException("sqsSrcUrl not set");
        var attrs = await clients.SQS().GetQueueAttributesAsync(new GetQueueAttributesRequest
        {
            QueueUrl = srcUrl,
            AttributeNames = ["RedrivePolicy"],
        });
        Assertions.True(attrs.Attributes.ContainsKey("RedrivePolicy"), "GetRedrivePolicy: RedrivePolicy not found");
        Assertions.True(attrs.Attributes["RedrivePolicy"].Contains("maxReceiveCount"), "GetRedrivePolicy: missing maxReceiveCount");
    }

    private async Task TeardownDlqAsync(TestContext context)
    {
        var srcUrl = context.GetString("sqsSrcUrl");
        if (!string.IsNullOrWhiteSpace(srcUrl))
        {
            try { await clients.SQS().DeleteQueueAsync(new DeleteQueueRequest { QueueUrl = srcUrl }); } catch { }
        }
        var dlqUrl = context.GetString("sqsDlqUrl");
        if (!string.IsNullOrWhiteSpace(dlqUrl))
        {
            try { await clients.SQS().DeleteQueueAsync(new DeleteQueueRequest { QueueUrl = dlqUrl }); } catch { }
        }
    }

    // ── sqs-fifo ──

    private async Task CreateFifoQueueAsync(TestContext context)
    {
        var name = $"{context.RunId}-sqs-fifo.fifo";
        var response = await clients.SQS().CreateQueueAsync(new CreateQueueRequest
        {
            QueueName = name,
            Attributes = new Dictionary<string, string> { ["FifoQueue"] = "true" },
        });
        var url = response.QueueUrl;
        Assertions.NotBlank(url, "CreateFifoQueue: url");
        context.Set("sqsFifoUrl", url);
    }

    private async Task SendFifoMessageAsync(TestContext context)
    {
        var url = context.GetString("sqsFifoUrl") ?? throw new InvalidOperationException("sqsFifoUrl not set");
        var response = await clients.SQS().SendMessageAsync(new SendMessageRequest
        {
            QueueUrl = url,
            MessageBody = "fifo-message",
            MessageGroupId = "test-group",
            MessageDeduplicationId = $"{context.RunId}-dedup-1",
        });
        Assertions.NotBlank(response.MessageId, "SendFifoMessage: MessageId");
    }

    private async Task ReceiveFifoMessageAsync(TestContext context)
    {
        var url = context.GetString("sqsFifoUrl") ?? throw new InvalidOperationException("sqsFifoUrl not set");
        List<Message> messages = new();
        for (var i = 0; i < 5; i++)
        {
            var response = await clients.SQS().ReceiveMessageAsync(new ReceiveMessageRequest
            {
                QueueUrl = url,
                MaxNumberOfMessages = 10,
                WaitTimeSeconds = 1,
            });
            messages = response.Messages;
            if (messages.Count > 0)
            {
                break;
            }
        }
        Assertions.True(messages.Count > 0, "ReceiveFifoMessage: expected at least 1 message");
    }

    private async Task TeardownFifoAsync(TestContext context)
    {
        var url = context.GetString("sqsFifoUrl");
        if (!string.IsNullOrWhiteSpace(url))
        {
            try { await clients.SQS().DeleteQueueAsync(new DeleteQueueRequest { QueueUrl = url }); } catch { }
        }
    }

    private static string RequireQueueUrl(TestContext context, string key)
    {
        return context.GetString(key) ?? throw new InvalidOperationException($"{key} not set");
    }
}

File: compat/suites/java-sdk/src/main/java/io/overcast/compat/groups/SqsGroup.java
package io.overcast.compat.groups;

import io.overcast.compat.clients.AwsClients;
import io.overcast.compat.harness.Assertions;
import io.overcast.compat.harness.TestContext;
import io.overcast.compat.harness.TestFn;
import software.amazon.awssdk.services.sqs.SqsClient;
import software.amazon.awssdk.services.sqs.model.*;

import java.util.List;
import java.util.Map;

/**
 * SQS compatibility test group.
 *
 * <p>Groups: sqs-messages, sqs-dlq, sqs-fifo. sqs-queues is a ported group, resolved
 * from compat/model/authored/sqs-queues.json by the scenario backend (#1903).
 */
public final class SqsGroup implements ServiceGroup {

    private final AwsClients clients;

    public SqsGroup(AwsClients clients) {
        this.clients = clients;
    }

    private SqsClient sqs() { return clients.sqs(); }

    @Override
    public Map<String, TestFn> impls() {
        return Map.ofEntries(
                Map.entry("sqs-messages:SendMessage",             this::sendMessage),
                Map.entry("sqs-messages:SendMessageBatch",        this::sendMessageBatch),
                Map.entry("sqs-messages:ReceiveMessage",          this::receiveMessage),
                Map.entry("sqs-messages:DeleteMessage",           this::deleteMessage),
                Map.entry("sqs-messages:ChangeMessageVisibility", this::changeMessageVisibility),
                Map.entry("sqs-messages:DeleteMessageBatch",      this::deleteMessageBatch),
                Map.entry("sqs-messages:PurgeQueue",              this::purgeQueue),
                Map.entry("sqs-dlq:CreateDLQ",                    this::createDlq),
                Map.entry("sqs-dlq:SetRedrivePolicy",             this::setRedrivePolicy),
                Map.entry("sqs-dlq:GetRedrivePolicy",             this::getRedrivePolicy),
                Map.entry("sqs-fifo:CreateFifoQueue",             this::createFifoQueue),
                Map.entry("sqs-fifo:SendFifoMessage",             this::sendFifoMessage),
                Map.entry("sqs-fifo:ReceiveFifoMessage",          this::receiveFifoMessage)
        );
    }

    @Override
    public Map<String, TestFn> setups() {
        return Map.ofEntries(
                Map.entry("sqs-messages", this::setupMessages),
                Map.entry("sqs-dlq",      this::setupDlq),
                Map.entry("sqs-fifo",     this::setupFifo)
        );
    }

    @Override
    public Map<String, TestFn> teardowns() {
        return Map.ofEntries(
                Map.entry("sqs-messages", ctx -> deleteQueueSilently(ctx.getString("sqsMsgQueueUrl"))),
                Map.entry("sqs-dlq",      this::teardownDlq),
                Map.entry("sqs-fifo",     ctx -> deleteQueueSilently(ctx.getString("sqsFifoUrl")))
        );
    }

    // ── sqs-messages ──────────────────────────────────────────────────────────

    private void setupMessages(TestContext ctx) throws Exception {
        String name = ctx.runId() + "-sqsmsg";
        var resp = sqs().createQueue(r -> r.queueName(name));
        ctx.set("sqsMsgQueueUrl", resp.queueUrl());
    }

    private void sendMessage(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsMsgQueueUrl");
        var resp = sqs().sendMessage(r -> r.queueUrl(url).messageBody("hello SQS"));
        Assertions.assertNotBlank(resp.messageId(), "SendMessage: messageId");
    }

    private void sendMessageBatch(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsMsgQueueUrl");
        var resp = sqs().sendMessageBatch(r -> r.queueUrl(url).entries(
                SendMessageBatchRequestEntry.builder().id("m1").messageBody("batch-1").build(),
                SendMessageBatchRequestEntry.builder().id("m2").messageBody("batch-2").build()
        ));
        Assertions.assertEquals(2, resp.successful().size(), "SendMessageBatch: expected 2 successful");
    }

    private void receiveMessage(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsMsgQueueUrl");
        var resp = sqs().receiveMessage(r -> r.queueUrl(url).maxNumberOfMessages(10).waitTimeSeconds(0));
        Assertions.assertNotEmpty(resp.messages(), "ReceiveMessage: no messages received");
        ctx.set("sqsReceiptHandle", resp.messages().get(0).receiptHandle());
    }

    private void deleteMessage(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsMsgQueueUrl");
        String rh  = ctx.getString("sqsReceiptHandle");
        if (rh != null) {
            sqs().deleteMessage(r -> r.queueUrl(url).receiptHandle(rh));
        }
    }

    private void changeMessageVisibility(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsMsgQueueUrl");
        // Send a fresh message to change visibility on.
        sqs().sendMessage(r -> r.queueUrl(url).messageBody("vis-test"));
        var recv = sqs().receiveMessage(r -> r.queueUrl(url).maxNumberOfMessages(1).waitTimeSeconds(1));
        Assertions.assertNotEmpty(recv.messages(), "ChangeMessageVisibility: no message to change visibility for");
        String rh = recv.messages().get(0).receiptHandle();
        sqs().changeMessageVisibility(r -> r.queueUrl(url).receiptHandle(rh).visibilityTimeout(0));
        // With the timeout set to 0 the message must be immediately receivable again.
        var recv2 = sqs().receiveMessage(r -> r.queueUrl(url).maxNumberOfMessages(1).waitTimeSeconds(1));
        Assertions.assertNotEmpty(recv2.messages(),
                "ChangeMessageVisibility: message not re-visible after setting timeout to 0");
    }

    private void deleteMessageBatch(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsMsgQueueUrl");
        sqs().sendMessage(r -> r.queueUrl(url).messageBody("batch-del-1"));
        sqs().sendMessage(r -> r.queueUrl(url).messageBody("batch-del-2"));
        var recv = sqs().receiveMessage(r -> r.queueUrl(url).maxNumberOfMessages(10).waitTimeSeconds(1));
        Assertions.assertNotEmpty(recv.messages(), "DeleteMessageBatch: no messages received to delete");
        var entries = recv.messages().stream()
                .map(m -> DeleteMessageBatchRequestEntry.builder()
                        .id(m.messageId()).receiptHandle(m.receiptHandle()).build())
                .toList();
        var resp = sqs().deleteMessageBatch(r -> r.queueUrl(url).entries(entries));
        Assertions.assertEquals(entries.size(), resp.successful().size(),
                "DeleteMessageBatch: not every entry deleted successfully (failed=" + resp.failed() + ")");
    }

    private void purgeQueue(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsMsgQueueUrl");
        sqs().purgeQueue(r -> r.queueUrl(url));
    }

    // ── sqs-dlq ────────────────────────────────────────────────────────────────

    private void setupDlq(TestContext ctx) throws Exception {
        String dlqName  = ctx.runId() + "-sqsdlq";
        String srcName  = ctx.runId() + "-sqssrc";
        var dlqResp     = sqs().createQueue(r -> r.queueName(dlqName));
        var dlqAttrs    = sqs().getQueueAttributes(r -> r
                .queueUrl(dlqResp.queueUrl())
                .attributeNames(QueueAttributeName.QUEUE_ARN));
        String dlqArn   = dlqAttrs.attributesAsStrings().get("QueueArn");
        var srcResp     = sqs().createQueue(r -> r.queueName(srcName));
        ctx.set("sqsDlqUrl", dlqResp.queueUrl());
        ctx.set("sqsDlqArn", dlqArn);
        ctx.set("sqsSrcUrl", srcResp.queueUrl());
    }

    private void teardownDlq(TestContext ctx) {
        deleteQueueSilently(ctx.getString("sqsDlqUrl"));
        deleteQueueSilently(ctx.getString("sqsSrcUrl"));
    }

    private void createDlq(TestContext ctx) {
        Assertions.assertNotBlank(ctx.getString("sqsDlqUrl"), "sqsDlqUrl");
    }

    private void setRedrivePolicy(TestContext ctx) throws Exception {
        String url    = ctx.getString("sqsSrcUrl");
        String dlqArn = ctx.getString("sqsDlqArn");
        String policy = "{\"deadLetterTargetArn\":\"" + dlqArn + "\",\"maxReceiveCount\":\"3\"}";
        sqs().setQueueAttributes(r -> r.queueUrl(url)
                .attributes(Map.of(QueueAttributeName.REDRIVE_POLICY, policy)));
    }

    private void getRedrivePolicy(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsSrcUrl");
        var resp = sqs().getQueueAttributes(r -> r
                .queueUrl(url).attributeNames(QueueAttributeName.REDRIVE_POLICY));
        Assertions.assertTrue(resp.attributesAsStrings().containsKey("RedrivePolicy"),
                "GetRedrivePolicy: RedrivePolicy attribute not present");
    }

    // ── sqs-fifo ───────────────────────────────────────────────────────────────

    private void setupFifo(TestContext ctx) throws Exception {
        String name = ctx.runId() + "-sqsfifo.fifo";
        var resp = sqs().createQueue(r -> r.queueName(name)
                .attributes(Map.of(QueueAttributeName.FIFO_QUEUE, "true")));
        ctx.set("sqsFifoUrl", resp.queueUrl());
    }

    private void createFifoQueue(TestContext ctx) {
        Assertions.assertNotBlank(ctx.getString("sqsFifoUrl"), "sqsFifoUrl");
    }

    private void sendFifoMessage(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsFifoUrl");
        var resp = sqs().sendMessage(r -> r.queueUrl(url)
                .messageBody("fifo-msg")
                .messageGroupId("grp1")
                .messageDeduplicationId("dedup1"));
        Assertions.assertNotBlank(resp.messageId(), "SendFifoMessage: messageId");
    }

    private void receiveFifoMessage(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsFifoUrl");
        var resp = sqs().receiveMessage(r -> r.queueUrl(url)
                .maxNumberOfMessages(1).waitTimeSeconds(2));
        Assertions.assertNotEmpty(resp.messages(), "ReceiveFifoMessage: no messages received");
        Assertions.assertEquals("fifo-msg", resp.messages().get(0).body(),
                "ReceiveFifoMessage: body mismatch");
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private void deleteQueueSilently(String url) {
        if (url == null) return;
        try { sqs().deleteQueue(r -> r.queueUrl(url)); } catch (Exception ignored) {}
    }
}
